### ISSUE TITLE
$_ incorrectly highlighted red

### ISSUE THREAD

**Issue by mxmilkb** [zsh-users/zsh-syntax-highlighting#619]
$_ (dollar underscore) is the last parameter from the previous line. I often (with autocd on) enter "$_" directly after a command to jump into the destination directory. However it's always highlighted red. Apologies if this is covered by another issue, I went through a number but couldn't find anything directly related. Thanks.

**Comment by danielshahaf** [comment]
Are literal directory names highlighted in green at command position when AUTO_CD is set?  AFAICS they aren't [can't test this myself right now], so you're essentially asking for AUTO_CD support to be added.  (Once literal directory names are handled, simple parameter expansions will DTRT too.)

Would you be interested in contributing this feature?

**Comment by danielshahaf** [comment]
By the way, `M-.` (Alt+.) is equivalent and has fewer keystrokes.

**Comment by mxmilkb** [comment]
Whoops, AUTO_CD paths are made purple (on my system) not green.

I'm afraid I don't have the understanding, time or energy to implement this.

Alt-. doesn't work for me, I guess either conflict with another zsh script or bad key binds. Thank you for the tip though.

**Comment by phy1729** [comment]
It's a bit more complicated than just supporting AUTO_CD. Currently `_` is not recognized as a parameter name https://github.com/zsh-users/zsh-syntax-highlighting/blob/d766243f7a2a920b01a405d8fd0b12cb18f6d5bc/highlighters/main/main-highlighter.zsh#L393 . The hard part is expanding `_` properly. We'd need to save the value of `_` and then special case the expansion code because `_` will have changed many times over once we've gotten to here https://github.com/zsh-users/zsh-syntax-highlighting/blob/d766243f7a2a920b01a405d8fd0b12cb18f6d5bc/highlighters/main/main-highlighter.zsh#L607

**Comment by danielshahaf** [comment]
That pattern _would_ match `$_` 's name, if we are able to compute the correct expansion, we might have another use-case for the envisioned `indeterminate` style — not just for `$_` but for any `[[ ${(t)foo} == *special* ]]` parameter at command position, unless whitelisted as an exception?

**Comment by danielshahaf** [comment]
See #617 for `indeterminate` (although that term isn't used there).

**Comment by danielshahaf** [comment]
My reading is two separate issues:

- [ ] Support AUTO_CD.
- [ ] Don't expand special parameters at command position.

**Comment by phy1729** [comment]
Indeed it does match `$_` oops. I agree in general special parameters should use a `indererminate` style; however, I think `$_` is useful enough to special case. There may be a cleaner way, but worst case we could save `${${(z)BUFFER}[-1]}` on `zle-line-finish`.

**Comment by danielshahaf** [comment]
> Indeed it does match `$_` oops. I agree in general special parameters 
> should use a `indererminate` style; however, I think `$_` is useful 
> enough to special case. There may be a cleaner way, but worst case we 
> could save `${${(z)BUFFER}[-1]}` on `zle-line-finish`.

And what guarantees us that value would be equal to the actual value of $_ that will be used when zsh executes the line?  What if there is another plugin after z-sy-h in the set of plugins and it writes to ${BUFFER} before the line is executed?

**Comment by phy1729** [comment]
Nothing. However, highlighting on the modified line would already be incorrect for those users (and we do document that we should be sourced last).

**Comment by danielshahaf** [comment]
Fair enough. Re 'cleaner way', I thought we could have zle expose the REPL-level value od `$_` through a new preset variable (akin to BUFFER, WIDGET, &c). WDYT?

**Comment by danielshahaf** [comment]
For the record, AUTO_CD support is in #669 (PR for 0.8.0) and parameters in command position are #674 (has WIP PR).  Support for `$_` specifically would require further work, as discussed in this issue.

**Comment by danielshahaf** [comment]
Cross-referencing #695.
